src/master.rs

```diff
@@ -517,9 +517,101 @@ impl Master {
         Ok(time)
     }
 
+    pub fn foe_read(&mut self, idx: SlavePos, name: &str) -> Result<Vec<u8>> {
+        let file_name = string_to_foe_name(name)?;
+        // FIXME: this is the same as in the c-implementation. Should read in chunks instead of a
+        // fixed size buffer. The ioctl-call in the master pre-allocates a 10000 byte buffer, so we
+        // do the same here.
+        const FOE_SIZE: usize = 10_000;
+        let mut buf: Vec<u8> = vec![0; FOE_SIZE];
+        let mut data = ec::ec_ioctl_slave_foe_t {
+            slave_position: idx.into(),
+            offset: 0,
+            buffer_size: FOE_SIZE as u64,
+            buffer: buf.as_mut_ptr(),
+            file_name,
+            ..Default::default()
+        };
+        ioctl!(self, ec::ioctl::SLAVE_FOE_READ, &mut data)?;
+
+        assert!(data.data_size <= FOE_SIZE as u64);
+        buf.resize(data.data_size as usize, 0x0);
+        Ok(buf)
+    }
+
+    pub fn foe_write(&mut self, idx: SlavePos, name: &str, data: &[u8]) -> Result<()> {
+        let file_name = string_to_foe_name(name)?;
+
+        let buffer = data.as_ptr() as *mut _;
+        let data = ec::ec_ioctl_slave_foe_t {
+            slave_position: idx.into(),
+            offset: 0,
+            buffer_size: data.len() as u64,
+            buffer,
+            file_name,
+            ..Default::default()
+        };
+        ioctl!(self, ec::ioctl::SLAVE_FOE_WRITE, &data)?;
+
+        Ok(())
+    }
+
     // XXX missing: write_idn, read_idn
 }
 
+fn string_to_foe_name(input: &str) -> Result<[std::os::raw::c_char; 32]> {
+    // construct the char buffer for the FoE name
+    let mut bytes = input.to_string().into_bytes();
+    if bytes.len() > 32 {
+        let e = io::Error::new(
+            io::ErrorKind::InvalidInput,
+            format!(
+                "FoE name can have a maximum length of 32, '{}' has {}",
+                input,
+                input.len()
+            ),
+        );
+        return Err(Error::Io(e));
+    }
+    bytes.resize(32, 0);
```

Comment:
My suggestion without allocations:
```
    // (length check here)
    let mut result = [0; 32];
    input.as_bytes().iter().zip(&mut result).for_each(|(&b, p)| *p = b as _);
    Ok(result)
```

Reply:
Nice one, thanks. 

And don't be sorry, I'm glad you're taking such a close look!